### ISSUE TITLE
Detecting years and year ranges

### ISSUE THREAD

**Issue by pbowyer** [facebook/duckling#564]
Duckling feels like magic, as it's so good at detecting dates and date ranges.

However there's one case I can't get it to do, which the demo at https://duckling.wit.ai/ does perfectly.

Years.

Specimen inputs:
```
1960            # Expected: 1960-01-01 00:00:00
1960-1961       # Expected: a range of 1960-01-01 00:00:00 to 1962-01-01 00:00:00
```
When I try these via the HTTP server, I get:
```
$ curl -XPOST http://0.0.0.0:8000/parse --data 'local=en_US&text="1960"&dims="[\"time\"]"'
[
  {
    "body": "1960\"",
    "start": 1,
    "value": {
      "second": 1960,
      "value": 1960,
      "type": "value",
      "unit": "second",
      "normalized": {
        "value": 1960,
        "unit": "second"
      }
    },
    "end": 6,
    "dim": "duration",
    "latent": false
  },
  {
    "body": "1960\"",
    "start": 1,
    "value": {
      "value": 1960,
      "type": "value",
      "unit": "inch"
    },
    "end": 6,
    "dim": "distance",
    "latent": false
  }
]


$ curl -XPOST http://0.0.0.0:8000/parse --data 'local=en_US&text="1960-1961"&dims="[\"time\"]"'                                                                                                                                                                                                                                                                            
[
  {
    "body": "1960-1961\"",
    "start": 1,
    "value": {
      "to": {
        "value": 1961,
        "unit": "inch"
      },
      "from": {
        "value": 1960,
        "unit": "inch"
      },
      "type": "interval"
    },
    "end": 11,
    "dim": "distance",
    "latent": false
  }
]
```
Seconds is a good guess and could be right; as would a guess of time for possible values (e.g. 1700 could mean 17:30). But in this case it's a year. And that doesn't appear as an option.

When I try these in debug, I get:
```
*Duckling.Debug> debug (makeLocale EN $ Just US) "1960" [Seal Time]
[]
*Duckling.Debug> debug (makeLocale EN Nothing) "1960-1961" [Seal Time]
[]
```

And https://duckling.wit.ai/#try-it-out? It gives
![image](https://user-images.githubusercontent.com/89852/103002905-b2c18480-4527-11eb-87b1-7417dc53da47.png)
![image](https://user-images.githubusercontent.com/89852/103002967-cd93f900-4527-11eb-8663-8c41d5366cd3.png)

**Comment by chessai** [comment]
Hi Peter!

You need to use the Duration dimension.

**Comment by pbowyer** [comment]
Hi @chessai, thanks for the help!

I've changed to the Duration dimension and unfortunately I see:
```
*Duckling.Debug> debug (makeLocale EN $ Just US) "1960" [Seal Duration]
[]
*Duckling.Debug> debug (makeLocale EN $ Just US) "1960-1961" [Seal Duration]
[]
```

and

```
$ curl -XPOST http://0.0.0.0:8000/parse --data 'local=en_US&text="1960"&dims="[\"time\", \"duration\"]"'
[
  {
    "body": "1960\"",
    "start": 1,
    "value": {
      "second": 1960,
      "value": 1960,
      "type": "value",
      "unit": "second",
      "normalized": {
        "value": 1960,
        "unit": "second"
      }
    },
    "end": 6,
    "dim": "duration",
    "latent": false
  },
  {
    "body": "1960\"",
    "start": 1,
    "value": {
      "value": 1960,
      "type": "value",
      "unit": "inch"
    },
    "end": 6,
    "dim": "distance",
    "latent": false
  }
]

[04:41 PM]-[vagrant@ngh]-[~/duckling]-[git master]
$ curl -XPOST http://0.0.0.0:8000/parse --data 'local=en_US&text="1960-1965"&dims="[\"time\", \"duration\"]"'
[
  {
    "body": "1960-1965\"",
    "start": 1,
    "value": {
      "to": {
        "value": 1965,
        "unit": "inch"
      },
      "from": {
        "value": 1960,
        "unit": "inch"
      },
      "type": "interval"
    },
    "end": 11,
    "dim": "distance",
    "latent": false
  }
]
```

**Comment by pbowyer** [comment]
I saw this commented out latent corpus entry https://github.com/facebook/duckling/blob/master/Duckling/Time/EN/Corpus.hs#L134-L136, so tried uncommenting it and enabling latent entries in the HTTP output. Either I failed at the latter or that rule isn't picked up by
```
$ curl -XPOST http://0.0.0.0:8000/parse --data 'locale=en_US&text="1954"&latent=True'
```

**Comment by chessai** [comment]
Firstly, I appreciate the very thorough issue. It makes understanding what's going on much easier.
Secondly, sorry for my quick responses - they weren't that helpful (holiday responses!)

Looks like I have a better understanding of what's going on now.

Currently duckling fails on something simple like "1960-1961":
```haskell
> debug (makeLocale EN Nothing) "1960-1961" [Seal Time]
[]
```
The pattern for that rule looks like this (regex simplified but maintains intent):
```haskell
-- <datetime> - <datetime> (interval)
pattern =
  [ Predicate isNotLatent
  , regex "\\-"
  , Predicate isNotLatent
  ]
```
If I change the pattern to this (note that now we accept latent times):
```haskell
pattern =
  [ dimension Time
  , regex "\\-"
  , dimension Time
  ]
```
Then the parse succeeds:
```haskell
> debug (makeLocale EN Nothing) "1960-1961" [Seal Time]
<datetime> - <datetime> (interval) (1960-1961)
-- year (latent) (1960)
-- -- integer (numeric) (1960)
-- -- -- regex (1960)
-- regex (-)
-- year (latent) (1961)
-- -- integer (numeric) (1961)
-- -- -- regex (1961)
...actual entity here...
```
But, now, phone numbers become time intervals, which is not what we want:
```
> debug (makeLocale EN Nothing) "650-333-6555" [Seal Time]
<datetime> - <datetime> (interval) (-333-6555)
-- year (latent) (-333)
-- -- negative numbers (-333)
-- -- -- regex (-)
-- -- -- integer (numeric) (333)
-- -- -- -- regex (333)
-- regex (-)
-- year (latent) (6555)
-- -- integer (numeric) (6555)
-- -- -- regex (6555)
...entities here...
```
Of course, if we add in Seal PhoneNumber in addition to Seal Time, the resolved value is just a phone number. But it still fails the negative corpus test, since clearly we don't want phone numbers to be recognised as time intervals.

Getting back to basics, the reason years are latent is that they're almost always false positives (most numeral strings could be considered a year). The best way I could see to approach this problem is to constrain the years to stuff that "makes sense". Firstly, we shouldn't accept negative numbers (almost everyone writes years as positive, even if it is _technically_ valid to do otherwise). Secondly, we could add some rules to see if it made sense for the years to be a range. Are they in the same millenium? century? Even so, doing this will still mean some ambiguity, but we can get a lot closer, and we certainly won't encounter ambiguity when parsing phone numbers in their standard forms. One obvious example showing a problem is the string "1800-2100". Does this mean "the datetime interval starting at the year 1800 and ending at the year 2100", or "the time interval 1800 hours to 2100 hours". We can do our best but there is always going to be some ambiguity there. 

To be clear, I'm in favour of doing better here - esp. since I'd like for duckling to be able to parse such simple things as "1960 - 1961". 

P.S. the one on wit is using an old version of duckling written in clojure. Occasionally people dig up discrepancies in their behaviour.

**Comment by chessai** [comment]
I have implemented a rule to catch latent year ranges such as yours. Here is what it looks like:

```haskell
-- In order to support latent year ranges, e.g. "1960 - 1961", we impose
-- the following constraints:
--   1. Neither year can be negative.
--   2. The first year must be less than the second year.
--   3. The years must be within the interval [1000,10000].
-- (1): We could try to allow negative years, but years in natural language are
--      almost never written as negative, so it's unnecessary complication.
-- (2): Year ranges in natural language are written
--      <earlier year> - <later year>. No need to derive the ordering from
--      something which is not likely a year range.
-- (3): Four+ digits prevents phone numbers (e.g. "333-444-5555") from
--      registering false positives.
--      In everyday language, people are more likely to mention years closer
--      to the present than very far in the past or future (closer times are
--      more relevant).
--      Of course, this means we do not have the ability to parse something like
--      "300 - 600" as a year interval. But, prior to implementing this
--      rule, we already did not have that ability.
--
-- These guidelines are not perfect, but they work. They can be iterated on and
-- improved going forward.
ruleIntervalYearLatent :: Rule
ruleIntervalYearLatent = Rule
  { name = "<year> (latent) - <year> (latent) (interval)"
  , pattern =
      let -- (1): neither year can be negative
          -- (3): years are in the interval [1e3,1e5]
          inRange = Predicate $ or . sequence [isIntegerBetween 1000 10000]
      in [ inRange
         , regex "\\-|to|th?ru|through|(un)?til(l)?"
         , inRange
         ]
  , prod = \case
      (t1:_:t2:_) -> do
        y1 <- getIntValue t1
        y2 <- getIntValue t2
        -- (2): the first year is less than the second year
        guard (y1 < y2)
        let td1 = mkLatent $ year y1
        let td2 = mkLatent $ year y2
        Token Time <$> interval TTime.Closed td1 td2
      _ -> Nothing
  }
```
It works:
```
> debug (makeLocale EN Nothing) "1960 to 1961" [Seal Time]
<year> (latent) - <year> (latent) (interval) (1960 to 1961)
-- integer (numeric) (1960)
-- -- regex (1960)
-- regex (to)
-- integer (numeric) (1961)
-- -- regex (1961)
[Entity {dim = "time", body = "1960 to 1961", value = RVal Time (TimeValue (IntervalValue (InstantValue {vValue = 1960-01-01 00:00:00 -0200, vGrain = Year},InstantValue {vValue = 1962-01-01 00:00:00 -0200, vGrain = Year})) [IntervalValue (InstantValue {vValue = 1960-01-01 00:00:00 -0200, vGrain = Year},InstantValue {vValue = 1962-01-01 00:00:00 -0200, vGrain = Year})] Nothing), start = 0, end = 12, latent = False, enode = Node {nodeRange = Range 0 12, token = Token Time TimeData{latent=False, grain=Year, form=Nothing, direction=Nothing, holiday=Nothing, hasTimezone=False}, children = [Node {nodeRange = Range 0 4, token = Token Numeral (NumeralData {value = 1960.0, grain = Nothing, multipliable = False, okForAnyTime = True}), children = [Node {nodeRange = Range 0 4, token = Token RegexMatch (GroupMatch ["1960"]), children = [], rule = Nothing}], rule = Just "integer (numeric)"},Node {nodeRange = Range 5 7, token = Token RegexMatch (GroupMatch ["",""]), children = [], rule = Nothing},Node {nodeRange = Range 8 12, token = Token Numeral (NumeralData {value = 1961.0, grain = Nothing, multipliable = False, okForAnyTime = True}), children = [Node {nodeRange = Range 8 12, token = Token RegexMatch (GroupMatch ["1961"]), children = [], rule = Nothing}], rule = Just "integer (numeric)"}], rule = Just "<year> (latent) - <year> (latent) (interval)"}}]
```
It is not in master yet.

As for latent years, they already work with no modifications:
```
> debugCustom testContext (Options {withLatent = True}) "1960" [Seal Time]
year (latent) (1960)
-- integer (numeric) (1960)
-- -- regex (1960)
[Entity {dim = "time", body = "1960", value = RVal Time (TimeValue (SimpleValue (InstantValue {vValue = 1960-01-01 00:00:00 -0200, vGrain = Year})) [SimpleValue (InstantValue {vValue = 1960-01-01 00:00:00 -0200, vGrain = Year})] Nothing), start = 0, end = 4, latent = True, enode = Node {nodeRange = Range 0 4, token = Token Time TimeData{latent=True, grain=Year, form=Nothing, direction=Nothing, holiday=Nothing, hasTimezone=False}, children = [Node {nodeRange = Range 0 4, token = Token Numeral (NumeralData {value = 1960.0, grain = Nothing, multipliable = False, okForAnyTime = True}), children = [Node {nodeRange = Range 0 4, token = Token RegexMatch (GroupMatch ["1960"]), children = [], rule = Nothing}], rule = Just "integer (numeric)"}], rule = Just "year (latent)"}}]
```

**Comment by chessai** [comment]
closed in 7193caafb982ae5f7762b82c41ca5b9f85f259a9
